src/rules/noImproperConstructorParamUsageRule.ts

```diff
@@ -0,0 +1,96 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Lint from "../lint";
+import * as ts from "typescript";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-improper-constructor-param-usage",
+        description: "Disallows constructor parameters to be accessed without the 'this.' prefix",
+        rationale: Lint.Utils.dedent`
+            This helps enforce consistancy.
+            When a constructor parameter is accessed without the 'this.' prefix, it is actually not acessing the same thing.
+            Example: 
+            class Foo {
+                constructor(public num: number) {
+                    this.num = 10;
+                    num = 5;  //this should be disallowed!
+                }
+            }
+            const f = new Foo();
+            alert(f.num);  // will display 10`,
+        optionsDescription: "Not configurable.",
+        options: null,
+        optionExamples: ["true"],
+        type: "functionality",
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING_FACTORY = (paramName: string) => {
+        return `The constructor parameter '${paramName}' should only be accessed with the 'this' keyword: 'this.${paramName}'`;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        const languageService = Lint.createLanguageService(sourceFile.fileName, sourceFile.getFullText());
+        return this.applyWithWalker(new NoImproperCtorParamUsageWalker(sourceFile, this.getOptions(), languageService));
+    }
+}
+
+export class NoImproperCtorParamUsageWalker extends Lint.RuleWalker {
+
+    private languageService: ts.LanguageService;
+
+    constructor(sourceFile: ts.SourceFile, options: Lint.IOptions, languageService: ts.LanguageService) {
+        super(sourceFile, options);
+        this.languageService = languageService;
+    }
+
+    public visitConstructorDeclaration(node: ts.ConstructorDeclaration) {
+        if (node.parameters && node.parameters.length > 0) {
+            const fileName = this.getSourceFile().fileName;
+            node.parameters.forEach((param: ts.ParameterDeclaration) => {
+                this.validateParam(param, fileName);
+            });
+        }
+        super.visitConstructorDeclaration(node);
+    }
+
+    private validateParam(param: ts.ParameterDeclaration, fileName: string) {
+        const highlights = this.languageService.getDocumentHighlights(fileName, param.name.getStart(), [fileName]);
```

Comment:
We only need to validate if the parameter is actually a parameter property, right? We check for that in another rule like this: https://github.com/palantir/tslint/blob/81c538c2cdc03bf319f5953bd573271420ff978a/src/rules/noConstructorVarsRule.ts#L48